agent/api/task.go

```diff
@@ -193,13 +193,23 @@ func (task *Task) updateTaskKnownStatus() (newStatus TaskStatus) {
 
 	// Set to a large 'impossible' status that can't be the min
 	earliestStatus := ContainerZombie
+	essentialContainerStopped := false
 	for _, cont := range task.Containers {
 		contKnownStatus := cont.GetKnownStatus()
+		if contKnownStatus == ContainerStopped && cont.Essential {
+			essentialContainerStopped = true
+		}
 		if contKnownStatus < earliestStatus {
 			earliestStatus = contKnownStatus
 		}
 	}
 
+	// If the essential container is stopped while other containers may be running
+	// don't update the task status until the other containers are stopped.
+	if earliestStatus == ContainerRunning && essentialContainerStopped {
```

Comment:
yes, you're right. This only prevent the task knownstatus moved to running, with this change it will go directly from pending to stopped(when all the other containers are stopped by the engine)